### ISSUE TITLE
Model.object and djangos @classproperties not working since 1.13.1

### ISSUE THREAD

**Issue by djbrown** [typeddjango/django-stubs#1285]
# Bug report

Hello :-) first of all thank you very much for releasing the new version 1.13.1 so quickly after I asked for it!
But now I am facing two new problems, which might be related so I'm posting them together.

## Steps to reproduce

See my separately crafted repo with minimal steps to reproduce:
https://github.com/djbrown/mypy-django-bugs

But basically it's:

* `pip install django mypy django-stubs==1.13.0`
* generate simple django project and include above bug snippets
* execute mypy check -> ✔
* `pip install django-stubs==1.13.1`
* execute mypy check -> ❌
* `pip install django-stubs==1.13.0`
* execute mypy check -> ✔

So it's definitely the new version..
I can also add a failing test or workflow so you don't have to reproduce manually.

<details>
<summary>[Collapsed: This part of the report was resolved]</summary>

## no attribute Model.objects
```python
from django.db.models import Model
def invalid(model: Model):
    model.objects
```
Results in `error: "Model" has no attribute "objects"  [attr-defined]`
I already tried to fix by enabling monkey-patching in settings.py or module via [django_stubs_ext](https://github.com/typeddjango/django-stubs/tree/master/django_stubs_ext) or use ModelBase or TypeVar instead of Model.
Because in https://github.com/typeddjango/django-stubs/pull/1150 the object property was moved from Model to ModelBase - but that doesn't work for me.

Maybe also related issues:
- https://github.com/typeddjango/django-stubs/issues/83
- https://github.com/typeddjango/django-stubs/issues/1068
  - https://github.com/typeddjango/django-stubs/issues/16

</details>

## cannot access classproperties
```python
from django.test import LiveServerTestCase
class ServerTestCase(LiveServerTestCase):
    def invalid(self, arg: str):
        self.live_server_url
```
Results in:
```
error: Argument 1 to "__get__" of "classproperty" has incompatible type "ServerTestCase"; expected "Optional[classproperty[Any]]"  [arg-type]
error: Argument 2 to "__get__" of "classproperty" has incompatible type "Type[ServerTestCase]"; expected "Type[classproperty[Any]]"  [arg-type]
```
I already tried to fix by enabling monkey-patching in settings.py or module via [django_stubs_ext](https://github.com/typeddjango/django-stubs/tree/master/django_stubs_ext), or access classproperty via class or super class.
Accessing normal class attributes works fine though.

## System information

- OS: ubuntu
- `python` version: 3.10.4
- `django` version: 4.1.4
- `mypy` version: 0.991
- `django-stubs` version: 1.13.1
- `django-stubs-ext` version: 0.7.0

**Comment by djbrown** [comment]
A possible workaround would be to to remove the type hints. That probably signals mypy to leave out the function.
I decided to comment-ignore the problematic lines until this is resolved.

**Comment by intgr** [comment]
> ```python
> from django.db.models import Model
> def invalid(model: Model):
>     model.objects
> ```
> Results in error: "Model" has no attribute "objects" [attr-defined]

This error is correct. `model: Model` tells mypy that the variable is an *instance* of Model.

Only Model classes have this attribute, not instances.

Try `def invalid(model: Type[Model]):`

**Comment by intgr** [comment]
As for the `classproperty` case, I am not sure what is happening there.

**Comment by djbrown** [comment]
ahhh thank you, that's correct! the Model.object problem is fixed now :-)

**Comment by dlesbre** [comment]
Hmm I've add a similar-ish problem.

```python
from typing import TypeVar
from django.db import models

T = TypeVar("T", bound=models.Model)

class MyQuerySet(models.QuerySet[T]):
    pass

class MyModelA(models.Model):
    object = MyQuerySet.as_manager()
```

With version 1.13.0 this passes fine, but with 1.13.1 mypy raises an error: 
```
error: Need type annotation for "object"  [var-annotated]
```

This minimal example can be solved by adding a type annotation:
```python
class MyModelA(models.Model):
    object: "models.Manager[MyModelA]" = MyQuerySet.as_manager()
```
However that wasn't necessary in the previous version. Moreover, my full code has two models mutually referencing each other and the type annotation leads to a new error I have no idea how to fix:
```python
from typing import TypeVar
from django.db import models

T = TypeVar("T", bound=models.Model)

class MyQuerySet(models.QuerySet[T]):
    pass

class MyModelA(models.Model):
    object: "models.Manager[MyModelA]" = MyQuerySet.as_manager()

    modelB = models.ForeignKey("MyModelB", on_delete=models.CASCADE)


class MyModelB(models.Model):
    modalA = models.ForeignKey(
        "MyModelA", null=True, blank=True, on_delete=models.SET_NULL
    )
```
which with mypy (and either version 1.13.0 or 1.13.1) gives a new error:
```
error: Couldn't resolve related manager for relation 'mymodela' (from myapp.models.MyModelA.myapp.MyModelA.modelB).  [django-manager-missing]

```

System information:

- OS: Kubuntu 22.10
- python 3.10.7
- django 4.1.4 
- mypy 0.991
- django-stubs-ext 0.7.0

**Comment by intgr** [comment]
@dlesbre Please open a separate issue, tracking multiple different problems in one issue gets messy.

**Comment by dlesbre** [comment]
> @dlesbre Please open a separate issue, tracking multiple different problems in one issue gets messy.

Done #1286

**Comment by djbrown** [comment]
The stub for [django.test.testcases.LiveServerTestCase](https://github.com/django/django/blob/9da2210f121085e998dc7cedb783bc4568390b92/django/test/testcases.py#L1812) looks currenty look like this:
https://github.com/typeddjango/django-stubs/blob/090aea6783a4fa45ab6419fb2c22c286d8a5852c/django-stubs/test/testcases.pyi#L256-L265

mypy complains about the accessed classproperties e.g. `live_server_url` or `allowed_host`,
but the normal class attributes can be used just fine e.g. `port`, `host` ...

I tried to set some other type hints instead of `cls: Any` (e.g. `type`, `type[Any]` or `type[Self]`) or use TypeVar or remove the annotation in the local testcases.pyi stub file, but only got slightly different mypy error: `... expected "Optional[classproperty[<nothing>]]"`

What kind of did work, was to replace `@classproperty` with `@classmethod` but then I had to add parenthesis to my actual source code (`live_server_url()`) and that doesn't match the django implementation.
The code wont actually work that way, so this is not a solution, but maybe someone can find a proper solution with that.
Maybe djangos custom [@classproperty](https://github.com/django/django/blob/0dd29209091280ccf34e07c9468746c396b7778e/django/utils/functional.py#L61) decorator has to be typed differently? It's not from the standard python library..

The [python typing docs](https://typing.readthedocs.io/en/latest/source/stubs.html#id22) say:
> Do not annotate self and cls in method definitions, except when referencing a type variable.

So I guess the `cls: type[Self]` annotation in other classmethods should be removed too, since they don't use type variables.

**Comment by djbrown** [comment]
What finally did work for me (based on my previous comments last concerns) was to just set the classproperties as simple attributes in the type stub without any decorators:
```python
class LiveServerTestCase(TransactionTestCase): 
    host: str 
    port: int 
    ...
    live_server_url: str
    allowed_host: str
```
I will create a pull request for that and hope you agree to this solution.
Alternatively someone will have to think about how to properly annotate djangos classproperties..

**Comment by intgr** [comment]
~~The `LiveServerTestCase` and `classproperty` classes don't seem to have changed between 0.13.0...1.13.1. Are you sure this regressed in 1.13.1?~~

Edit: Oh, sorry, I had some old commit checked out. I think I see the bug now.

**Comment by intgr** [comment]
Indeed, classproperty was broken, fixed in #1287.

**Comment by djbrown** [comment]
thanks :) I reviewed your PR and applied the patch locally. that fixes the problem too - and more correctly :)
